### PR TITLE
perf: improve autocomplete performance

### PR DESCRIPTION
**Pull request by AyushAgrawal-A2**
for #1556 
This PR is meant for trial and discussion, may not merge

Problems causing #1556 
All workers are busy due to large number of operations. After waiting for some time app becomes responsive again. To see this visually, create a cell with some formatting and sheet fill. Drag it, via autocomplete, to a large selection. Water falling format updates can be seen progressing.

This is happening because each formatting update for every cell is applied in a separate operation. This leads to:
- Large number of inter worker message which take a lot of time.
- Sending individual cell update for each operation, cause unnecessary render updates.
- Use of `transaction.reverse_operations.insert(0, ...` which is O(n^2) for n operations.

In this PR, I am trying to batch client updates. Sending update for the entire affected `sheet_rect` is faster than filter and sending only changes for each individual cell.

Current performance in this PR is a lot faster than before but still slower than google sheets / excel web. More optimisations can be done.
Also this can improve performance of copy paste as well. Not implemented in this PR.

### CODE REVIEW

**Comment by davidfig** [comment]
Awesome. This was definitely on my list as well. I thought of two approaches for this problem: (1) your approach to batch the client calls to avoid rerendering hashes multiple times; and (2) rewriting the Operation::SetCellValues to be more efficient and accept something other than CellValues. The problem with CellValues is that it has to be a rectangle, and autocomplete is always a rectangle with a corner removed.

What are your thoughts on continued improvement?

**Comment by AyushAgrawal-A2** [comment]
@davidfig 
I am already trying to figure out -
- Refactoring to Operation::SetCellFormatsSelection 
- How to do batch client update for multiplayer, currently only operations are sent to multiplayer.

Current bottleneck is in `finalize_transaction` step, after sending batch client update. Size of reverse_operations vector is huge. Processing it and sending all those transactions is taking >90% of time currently.

hash boundaries bug is fixed in #1526

**Comment by davidfig** [comment]
Cool. Great on the hash boundaries bug. 

I'm not sure if we need to batch the client updates for multiplayer. We've been handling multiplayer via transactions, and I'm reluctant to change that.

Ideally, we should reduce the number of the operations.

**Comment by AyushAgrawal-A2** [comment]
@davidfig 
I am also reluctant to change multiplayer, a lot of code I don't understand yet.
But I don't know how else to solve water falling client update and large number of worker messages for multiplayer.

Reducing operations is priority, I am trying to do that too but even if there are a ton of operations multiplayer should be able to batch update same as host client is doing.

**Comment by davidfig** [comment]
If the batching works at the level of the transaction, then mp should work fine. It runs the ops via shared transactions.

**Comment by AyushAgrawal-A2** [comment]
@davidfig 
alot faster now :man_dancing:

**Comment by davidkircos** [comment]
Yeah, wow, this is way faster! Is this PR ready for review?

**Comment by AyushAgrawal-A2** [comment]
@davidkircos 
I would like to add some test if we are to merge this PR

Also, I am not able to figure out how to batch client update for multiplayer. Water falling is still present for MP. Can be done in another PR.

Otherwise, ready for user testing

**Comment by AyushAgrawal-A2** [comment]
@davidfig 
switch to `SetCellFormatsSelection`
faster than excel web now, I am satisfied with the performance

**Comment by davidfig** [comment]
Awesome. Can you also increase test coverage? Thanks!

**Comment by AyushAgrawal-A2** [comment]
> Awesome. Can you also increase test coverage? Thanks!

Yup, on it

**Comment by AyushAgrawal-A2** [comment]
I wanted to add `expect_js_call_count` in test for `batch_client_update` but this required adding `#[parallel]` macro to all tests.
I have already done that in #1526. Can be done after merging it.

**Comment by davidfig** [comment]
Can you add the batch_client_update_rect to receive_multiplayer#received_transction? Wouldn't that add this functionality to mp, or is there more that needs to happen?

File: quadratic-core/src/controller/execution/execute_operation/mod.rs [hunk 1]
```diff
@@ -48,7 +48,10 @@ impl GridController {
                 }
             }
 
-            if cfg!(target_family = "wasm") && !transaction.is_server() {
+            if (cfg!(target_family = "wasm") || cfg!(test))
+                && !transaction.is_server()
+                && transaction.batch_client_update_rect.is_none()
```
Comment (davidfig) [inline quadratic-core/src/controller/execution/execute_operation/mod.rs:53]: Why are you checking the batch here?

**Review by davidfig — CHANGES_REQUESTED** [review]

File: quadratic-core/src/controller/execution/execute_operation/mod.rs (hunk 1 above)
Comment (AyushAgrawal-A2) [reply quadratic-core/src/controller/execution/execute_operation/mod.rs:53]: to solve too many worker messages for transactions having large number of operations
worker message are very slow

**Comment by davidfig** [comment]
Cool. If you can make it dynamic, then mp should just work. We can be naive and use a Bounds struct that allows you to increase its size based on what the ops are doing.

**Comment by AyushAgrawal-A2** [comment]
I hate this bot :angry: 
image

image

**Comment by AyushAgrawal-A2** [comment]
> Cool. If you can make it dynamic, then mp should just work. We can be naive and use a Bounds struct that allows you to increase its size based on what the ops are doing.

yup, I figured this much. we still need a way to tell mp than it needs to batch operations and not send individual updated.
maybe based on operations vector size, I am sure now
current bottleneck is `send_transactions` which should be solve by #1607

**Comment by davidfig** [comment]
> we still need a way to tell mp than it needs to batch operations and not send individual updated

If it's dynamically updated in a more complicated Batch struct, then we'll know exactly what needs to be sent to the client based on the execute operation (which should update the Batch struct as it executes).

**Comment by AyushAgrawal-A2** [comment]
> > we still need a way to tell mp than it needs to batch operations and not send individual updated
> 
> If it's dynamically updated in a more complicated Batch struct, then we'll know exactly what needs to be sent to the client based on the execute operation (which should update the Batch struct as it executes).

I disagree actually, having fine grained updates was the biggest issue that this PR solve. This will again lead to multiple re-renders and a ton of worker messages depending on the complexity of Batch struct.

My idea behind batch update is to not figure out finer updates, but instead to send a larger rect re-render to avoid 
- multiple rerenders of  same hash 
- number of worker messages

Worker messages is the slowest step here, which I am trying to optimise

**Comment by davidfig** [comment]
Hmm... I'm not sure I follow. My suggestion is to create a Batch struct that tracks all changes made by operations. Then in finalize_transaction, we use the Batch struct to send the necessary messages.

To avoid too many messages, you can combine the rects before sending the messages. For example, you collect a set of cell hashes that have changed based on the rects, then you only send those cell hashes (so each cell hash is only sent once).

You end up with the correct number of messages this way without having to define the rect before the transaction is executed. Does that make sense?

**Comment by AyushAgrawal-A2** [comment]
@davidfig 
Batch update part is not doing much now after moving to 'SetCellFormatsSelection'. Earlier each cell had 10 operations for format update. Now we have just one operation for all format updates for the entire rect.
I should have done this before playing with other code.

As we have just ~2 operations now re-renders and worker messages are almost same as what we would achieve with a Batch struct.

I have disable it and push it here. Please try it on Vercel preview. It is pretty fast. I tried autocomplete for a very large rect `1x4` to `10000 x 8`. Add lot of different formatting in source rect to stress the operation. It looks OK to me.

I think we can remove & postpone Batch update logic.
I get what you are saying but I am not able to think of a scenario which I can implement and compare practically.

**Comment by davidfig** [comment]
Sounds good. We'll likely revisit the batching down the road.

**Comment by AyushAgrawal-A2** [comment]
> Sounds good. We'll likely revisit the batching down the road.

Great... Thanks for working this out with me. This PR was meant to do exactly this :smiley: 

Also removing batch_update logic might satisfy the insatiable codecov bot :stuck_out_tongue_closed_eyes:

**Pull request merged**